map.go

```diff
@@ -0,0 +1,124 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+)
+
+// Map implements the basic map function.
+func Map[In, Out any](arr []In, fn func(In) Out) []Out {
+	out := make([]Out, 0, len(arr))
+	for _, elem := range arr {
+		out = append(out, fn(elem))
+	}
+	return out
+}
+
+// MapError is the same as `Map` but for functions that might return an error.
+func MapError[In, Out any](arr []In, fn func(In) (Out, error)) ([]Out, error) {
+	out := make([]Out, 0, len(arr))
+	for i, elem := range arr {
+		mapped, err := fn(elem)
+		if err != nil {
+			return nil, fmt.Errorf("MapError got an error in index %d, value %v: %w", i, elem, err)
+		}
+		out = append(out, mapped)
+	}
+	return out, nil
+}
+
+// MapContextError is the same as `MapError` but with a context.
+func MapContextError[In, Out any](
+	ctx context.Context,
+	arr []In,
+	fn func(context.Context, In) (Out, error),
+) ([]Out, error) {
+	out := make([]Out, 0, len(arr))
+	for i, elem := range arr {
+		mapped, err := fn(ctx, elem)
+		if err != nil {
+			return nil, fmt.Errorf("MapContextError got an error in index %d, value %v: %w", i, elem, err)
+		}
+		out = append(out, mapped)
+	}
+	return out, nil
+}
+
+// MapConcurrent is the same as `MapContextError` but applied with concurrency.
+// In spite of concurency, order is guaranteed.
+func MapConcurrent[In, Out any](
```

Comment:
I wonder if we should call this func MapConcurrentToResults to differentiate it from the next MapConcurrentError